src/troubleshooting/content.md

```diff
@@ -0,0 +1,641 @@
+
+When you wish to deploy an application in Kubernetes, you usually define three components:
+
+- a Deployment — which is a recipe for creating copies of your application called Pods
+- a Service — an internal load balancer that routes the traffic to Pods
+- an Ingress — a description of how the traffic should flow from outside the cluster to your Service.
+
+Here's a quick visual recap.
+
+```slideshow
+{
+  "description": "Kubernetes core concepts: Ingress, Service, Pod and Deployment.",
+  "slides": [
+    {
+      "image": "core-concepts-1.svg",
+      "description": "In Kubernetes your applications are exposed through two layers of load balancers: internal and external."
+    },
+    {
+      "image": "core-concepts-2.svg",
+      "description": "The internal load balancer is called Service, whereas the external one is called Ingress."
+    },
+    {
+      "image": "core-concepts-3.svg",
+      "description": "Pods are not deployed directly. Instead, the Deployment creates the Pods and whatches over them."
+    }
+  ]
+}
+```
+
+Assuming you wish to deploy a simple _Hello World_ application, the YAML for such application should look similar to this:
+
+```yaml|highlight=2,23,34|title=hello-world.yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: app
+  labels:
+    track: canary
+spec:
+  selector:
+    matchLabels:
+      name: app
+  template:
+    metadata:
+      labels:
+        name: app
+    spec:
+      containers:
+      - name: app
+        image: learnk8s/app:1.0.0
+        ports:
+        - containerPort: 8080
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: app
+spec:
+  ports:
+  - port: 80
+    targetPort: 8080
+  selector:
+    name: app
+---
+apiVersion: networking.k8s.io/v1beta1
+kind: Ingress
+metadata:
+  name: app
+spec:
+  rules:
+  - http:
+    paths:
+    - backend:
+        serviceName: app
+        servicePort: 80
+      path: /
+```
+
+The definition is quite long, and it's easy to overlook how the components relate to each other.
+
+For example:
+
+- When should you use port 80 and when port 8080?
+- Should you create a new port for every Service so that they don't clash?
+- Do label names matter? Should it be the same everywhere?
+
+Before focusing on the debugging, let's recap how the three components link to each other.
+
+Let's start with Deployment and Service.
+
+## Connecting Deployment and Service
+
+The surprising news is that Service and Deployment aren't connected at all.
+
+Instead, the Service points to the Pods directly and skips the Deployment altogether.
+
+So what you should pay attention to is how the Pods and the Service are related to each other.
+
+You should remember three things:
+
+1. The Service selector should match the label of the Pod
+1. The Service `targetPort` should match the `containerPort` of the container inside the Pod
+1. The Service `port` can be any number. Multiple Service can use the same port because they have different IP addresses assigned.
+
```

Comment:
General comment: did you know that you can name ports, so you can write these 3 YAML files in a way that everything is connected by NAMES and not 8080s, so you can change it without much confusion.  This is why we have containerPort in Pod, so you can assign a string name.  If you are not using that feature, then the whole containerPort thingie is a little bit ridiculous and unnecessary, the important thing is that you then listen from your app on that port.

The article in the current format a little bit suggest that kubernetes can somehow communicate which port the app should listen to (interesting tidbit: this was the solution that Google used internally with Borg, so that is a possible design).

Reply:
I'm aware of the names but I always thought that (as a beginner) you lose track of what is what. Maybe I'm wrong and I should revisit. I don't think I want to revisit in this blog post though.

> The article in the current format a little bit suggest that kubernetes can somehow communicate which port the app should listen 

I think this is a sort of "lie" that I can live with. It's easy to digest and believe.
The fact that is not true isn't a shocking revelation as it doesn't change it much. It's much harder to explain though.
I think it's a reasonable compromise.